### ISSUE TITLE
[CAL-4864] Recurring event: No available users found error

### ISSUE THREAD

**Issue by CarinaWolli** [calcom/cal.com#18072]
For recurring events, we want to check the availability for the first two slots only. If other slots are not available the booking should still go through. 

Currently, it throws an error when the second slot is not available: 

![Screenshot 2024-12-09 at 10.10.47 AM.png](https://uploads.linear.app/e86bf957-d82f-465e-b205-135559f4b623/73a0e315-dbf6-403c-9a9e-bcdd2704e39b/7eb3f6d8-6a19-4837-af21-45fdc939c130?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiL2U4NmJmOTU3LWQ4MmYtNDY1ZS1iMjA1LTEzNTU1OWY0YjYyMy83M2EwZTMxNS1kYmY2LTQwM2MtOWE5ZS1iY2RkMjcwNGUzOWIvN2ViM2Y2ZDgtNmExOS00ODM3LWFmMjEtNDVmZGM5MzljMTMwIiwiaWF0IjoxNzMzNzU3MTI5LCJleHAiOjE3NjUyOTMxMjl9.rzlKsUQi657PhtW5CPi36xwqrVmk27vgA-MUjObfe4E)

We need to make sure to only load slots when the first two are available. We never want to throw this error. 

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-4864](https://linear.app/calcom/issue/CAL-4864/recurring-event-no-available-users-found-error)</sub>

**Comment by sanchitttt** [comment]
# Issue Replication Scenario for Recurring Events Booking

### Scenario to Replicate the Bug in the Production Environment

1. **Availability Setup:**
   - Set availability for the following dates: 
     - **December 11th**: 10 AM - 11 AM
     - **December 18th**: 10 AM - 11 AM
     - **December 25th**: 10 AM - 11 AM

2. **Create Recurring Event:**
   - Set up a recurring event that repeats every week for **6 occurrences**.

3. **Booking Scenario:**
   - **User A** books the slot on **December 18th** (1 occurrence).
   - **User B** books the slot on **December 11th** (2 occurrences).

### Expected Behavior:

- **December 11th's slot** should be successfully booked for **User B**.
- **December 18th's slot** should be ignored in terms of checking availability, and the booking should still be processed for **User A**, as only the first two slots (December 11th and 18th) are being checked for availability.
- The error should **not** be thrown when booking the second slot (December 18th) as long as the first slot is available.

### Actual Behavior:

- An error is thrown when attempting to book for **December 18th** (second slot), even though the first slot (**December 11th**) is available.

**Current Date in Example**: **December 10th**

---

### Clarification:

Is my interpretation correct?

Or should **User B** get both **December 11th** and **December 25th** instead of just December 11th?

**Comment by CarinaWolli** [comment]
> Expected Behavior:
> December 11th's slot should be successfully booked for User B.
> December 18th's slot should be ignored in terms of checking availability, and the booking should still be processed for User A, as only the first two slots (December 11th and 18th) are being checked for availability.
> The error should not be thrown when booking the second slot (December 18th) as long as the first slot is available.

No. After December 18th is booked, then Dec 11th should not even be shown as available anymore if we are trying to book more occurrences. Currently, we check on booking confirmation the availability of the first two slots. We need to do the same for when loading slots, so also check for first two slots when loading slots 


> Actual Behavior:
> An error is thrown for User B when attempting to book for December 11th (first slot), since the second slot (December 18th) is unavailable.

Correct

**Comment by sanchitttt** [comment]
> > Expected Behavior:
> > December 11th's slot should be successfully booked for User B.
> > December 18th's slot should be ignored in terms of checking availability, and the booking should still be processed for User A, as only the first two slots (December 11th and 18th) are being checked for availability.
> > The error should not be thrown when booking the second slot (December 18th) as long as the first slot is available.
> 
> No. After December 18th is booked, then Dec 11th should not even be shown as available anymore if we are trying to book more occurrences. Currently, we check on booking confirmation the availability of the first two slots. We need to do the same for when loading slots, so also check for first two slots when loading slots

By that logic, the user shouldn't see any slots as available right? Since for my scenario if I only have Dec 11, 18 and 25 as availability. 18 is already booked. 11 cant be visible since it's next slot (ie 18) is already booked. And 25 cannot be booked since there is not availability after 25.

Can this statement be considered correct - "**Every availability must have a consecutive availability for it to show for recurring events**"

Also, if the user manually updates the occurences input box to 1, then will 11 and 25 be visible? [Edge case I think]

**Comment by rafarc21** [comment]
We are also struggling a lot with this issue. Would be extremely grateful to see it fixed. Happy to contribute if needed.

**Comment by reubenleemiller** [comment]
Why has this not been made more of a priority, this has been causing many struggles for us as well, I really hope this gets resolved ASAP, because it is just ridiculous that the recurring event functionality was even advertised or placed in the app if it doesn't display the correct time slots

**Comment by CarinaWolli** [comment]
We are aware that recurring events have some issues, which is why this feature is still marked as experimental. Due to performance concerns, this is quite a complex feature, so we are still looking into how to get it right without introducing long loading times.

![Image](https://github.com/user-attachments/assets/dfa86b5d-e1f1-410c-93ba-14df4caf7499)
